src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1082,137 +1079,179 @@ public bool TakeStartLine(ReadableBuffer buffer, out ReadCursor consumed, out Re
 
             var needDecode = false;
             var pathBegin = methodEnd + 1;
-            var pathToEndSpan = span.Slice(pathBegin, span.Length - pathBegin);
-            pathBegin = 0;
+            var spaceIndex = -1;
+            var questionMarkIndex = -1;
+            var percentageIndex = -1;
+            var pathEnd = -1;
 
-            // TODO: IndexOfAny
-            var spaceIndex = pathToEndSpan.IndexOf(ByteSpace);
-            var questionMarkIndex = pathToEndSpan.IndexOf(ByteQuestionMark);
-            var percentageIndex = pathToEndSpan.IndexOf(BytePercentage);
+            // TODO: State machineify
 
-            var pathEnd = MinNonZero(spaceIndex, questionMarkIndex, percentageIndex);
 
-            if (spaceIndex == -1 && questionMarkIndex == -1 && percentageIndex == -1)
+            fixed (byte* data = &span.DangerousGetPinnableReference())
             {
-                RejectRequestLine(start, end);
-            }
-            else if (percentageIndex != -1)
-            {
-                needDecode = true;
-
-                pathEnd = MinNonZero(spaceIndex, questionMarkIndex);
-                if (questionMarkIndex == -1 && spaceIndex == -1)
+                int length = span.Length;
+                for (int i = pathBegin; i < length; i++)
                 {
-                    RejectRequestLine(start, end);
+                    var ch = data[i];
+                    if (spaceIndex == -1 && ch == ByteSpace)
+                    {
+                        if (pathEnd == -1)
+                        {
+                            pathEnd = i;
+                        }
+
+                        spaceIndex = i;
+                    }
+
+                    if (questionMarkIndex == -1 && ch == ByteQuestionMark)
+                    {
+                        if (pathEnd == -1)
+                        {
+                            pathEnd = i;
+                        }
+
+                        questionMarkIndex = i;
+                    }
+
+                    if (percentageIndex == -1 && ch == BytePercentage)
+                    {
+                        if (pathEnd == -1)
+                        {
+                            pathEnd = i;
+                        }
+
+                        percentageIndex = i;
+                    }
                 }
```

Comment:
Try replace loop between lines 1093-1125 for code below.
I tested this piece of code in console application and it's seems to be 40% faster.
`for (int i = pathBegin; i < length; i++)
{
    var ch = data[i];

    switch (ch)
    {
        case ByteSpace:
            if (spaceIndex == -1)
            {
                if (pathEnd == -1)
                {
                    pathEnd = i;
                }

                spaceIndex = i;

                if (questionMarkIndex != -1 && percentageIndex != -1)
                {
                    goto loopExit;
                }
            }
            break;
        case ByteQuestionMark:
            if (questionMarkIndex == -1)
            {
                if (pathEnd == -1)
                {
                    pathEnd = i;
                }

                questionMarkIndex = i;

                if (spaceIndex != -1 && percentageIndex != -1)
                {
                    goto loopExit;
                }
            }
            break;
        case BytePercentage:
            if (percentageIndex == -1)
            {
                if (pathEnd == -1)
                {
                    pathEnd = i;
                }

                percentageIndex = i;

                if (spaceIndex != -1 && questionMarkIndex != -1)
                {
                    goto loopExit;
                }
            }
            break;
    }
}

loopExit:`